src/physlr-molecules.cc

```diff
@@ -333,100 +545,281 @@ biconnectedComponents(
+	// Strategy: row-normalize then square the matrix.
+
+	int n = adj_mat.size();
+	std::vector<double> temp(n, 0.0);
+	std::vector<std::vector<double>> normalized(n, temp);
+	double row_sum = 0;
+
+	adjacencyMatrix_t::iterator row_i;
+
+	auto normalized_row_i = normalized.begin();
+	for (row_i = adj_mat.begin(); row_i != adj_mat.end(); ++row_i, ++normalized_row_i) {
+		row_sum = 0;
+		auto first = row_i->begin();
+		auto last = row_i->end();
+		while (first != last) {
+			row_sum += (*first) * (*first);
+			++first;
+		}
 
-			for (uint64_t j = 0; j < length; j++) {
-				binned_neighbours[counter_new].insert(*elementIt);
-				++elementIt;
+		first = row_i->begin();
+		auto first_normalized = normalized_row_i->begin();
+		while (first != last) {
+			if (row_sum > 0) {
+				*first_normalized = *first / sqrt(1.0 * row_sum);
+			} else {
+				*first_normalized = 0;
 			}
-			counter_new++;
+			++first;
+			++first_normalized;
 		}
 	}
+	cosimilarity = square_matrix_ikj(normalized);
 }
 
-template<class Neighbours_Type>
 void
-bin_neighbours(
-    Neighbours_Type neighbours,
-    componentToVertexSet_t& binned_neighbours,
-    uint64_t bin_size = 50)
+community_detection_cosine_similarity_core(
+    graph_t& subgraph,
+    componentToVertexSet_t& componentToVertexSet,
+    bool squaring = true,
+    double threshold = 0.5)
 {
-	// //   Randomly split the set of vertices (neighbours) into bins
+	// Detect communities using cosine similarity of vertices
 
-	componentToVertexSet_t compToVertset(1, vertexSet_t(neighbours.first, neighbours.second));
-	if (compToVertset[0].size() > bin_size) {
-		bin_components(compToVertset, binned_neighbours, bin_size);
+	// 0- Map indices and vertex names
+
+	vertexToIndex_t vertexToIndex;
+	uint64_t subgraph_size = boost::num_vertices(subgraph);
+	vertexToIndex.reserve(subgraph_size);
+
+	if (subgraph_size < 10) {
+		// Do nothing on subgraphs smaller than a certain size
+		threshold = 0;
+	}
+
+	adjacencyMatrix_t adj_mat(convert_adj_list_adj_mat(subgraph, vertexToIndex));
+	indexToVertex_t indexToVertex = inverse_map(vertexToIndex);
+
+	// 1- Calculate the cosine similarity:
+
+	int size_adj_mat = adj_mat.size();
+	std::vector<double> tempVector(size_adj_mat, 0);
+	std::vector<std::vector<double>> cosSimilarity2d(size_adj_mat, tempVector);
+
+	if (squaring) {
+		calculate_cosine_similarity_2d(square_matrix_ikj(adj_mat, true), cosSimilarity2d);
 	} else {
-		binned_neighbours = compToVertset;
+		calculate_cosine_similarity_2d(adj_mat, cosSimilarity2d);
+	}
+
+	// 2- Determine the threshold:
+	// not implemented yet; uses a predefined universal threshold.
+
+	threshold = threshold;
+
+	// 3- Filter out edges:
+
+	for (uint64_t i = 0; i < adj_mat.size(); i++) {
+		for (uint64_t j = i + 1; j < adj_mat.size(); j++) {
+			if (cosSimilarity2d[i][j] < threshold) {
+				adj_mat[i][j] = 0;
+				adj_mat[j][i] = 0;
+			}
+		}
+	}
+
+	// 4- Detect Communities (find connected components - DFS)
```

Comment:
factor the connected components out in my opinion.

Reply:
Good idea, will do that.